presto-delta/src/main/java/com/facebook/presto/delta/DeltaClient.java

```diff
@@ -0,0 +1,188 @@
+     * @return If the table is found return {@link DeltaTable}.
+     */
+    public Optional<DeltaTable> getTable(
+            ConnectorSession session,
+            SchemaTableName schemaTableName,
+            String tableLocation,
+            Optional<Long> snapshotId,
+            Optional<Long> snapshotAsOfTimestampMillis)
+    {
+        Optional<DeltaLog> deltaLog = loadDeltaTableLog(session, new Path(tableLocation));
+        if (!deltaLog.isPresent()) {
+            return Optional.empty();
+        }
+
+        // Fetch the snapshot info for given snapshot version. If no snapshot version is given, get the latest snapshot info.
+        // Lock the snapshot version here and use it later in the rest of the query (such as fetching file list etc.).
+        // If we don't lock the snapshot version here, the query may end up with schema from one version and data files from another
+        // version when the underlying delta table is changing while the query is running.
+        Snapshot snapshot;
+        if (snapshotId.isPresent()) {
+            try {
+                snapshot = deltaLog.get().getSnapshotForVersionAsOf(snapshotId.get());
+            }
+            catch (IllegalArgumentException iae) {
+                throw new PrestoException(
+                        NOT_FOUND,
+                        format("Snapshot version %d does not exist in Delta table '%s'.", snapshotId.get(), schemaTableName),
+                        iae);
+            }
+        }
+        else if (snapshotAsOfTimestampMillis.isPresent()) {
+            try {
+                snapshot = deltaLog.get().getSnapshotForTimestampAsOf(snapshotAsOfTimestampMillis.get());
+            }
+            catch (IllegalArgumentException iae) {
+                throw new PrestoException(
+                        NOT_FOUND,
+                        format(
+                                "There is no snapshot exists in Delta table '%s' that is created on or before '%s'",
+                                schemaTableName,
+                                Instant.ofEpochMilli(snapshotAsOfTimestampMillis.get())),
+                        iae);
+            }
+        }
+        else {
+            snapshot = deltaLog.get().snapshot();
+        }
+
+        Metadata metadata = snapshot.getMetadata();
+        String format = metadata.getFormat().getProvider();
+        if (!"parquet".equalsIgnoreCase(format)) {
+            throw new PrestoException(DELTA_UNSUPPORTED_DATA_FORMAT,
+                    format("Delta table %s has unsupported data format: %s. Currently only Parquet data format is supported", schemaTableName, format));
+        }
+
+        return Optional.of(new DeltaTable(
+                schemaTableName.getSchemaName(),
+                schemaTableName.getTableName(),
+                tableLocation,
+                Optional.of(snapshot.getVersion()), // lock the snapshot version
```

Comment:
I will be addressing this in a next path. I am also planning to add a Guava cache so that we don't load the `DeltaLog` again when listing the files in the `Snapshot`.